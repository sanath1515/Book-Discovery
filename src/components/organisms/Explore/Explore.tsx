import React, { FormEvent, useState } from "react";
import { withStyles } from "@material-ui/core/styles";
import { Button, makeStyles } from "@material-ui/core";
import Menu from "@material-ui/core/Menu";
import ExploreTab, { Category } from "../../molecules/ExploreTab/index";
import { COLORS } from "../../../theme/constants";
import ArrowDropDownOutlinedIcon from "@material-ui/icons/ArrowDropDownOutlined";
import ArrowDropUpOutlinedIcon from "@material-ui/icons/ArrowDropUpOutlined";
import theme from "../../../theme/theme";
import { items } from "../../molecules/ExploreTab/constants";
import { useDispatch } from "react-redux";
import { useHistory } from "react-router";

type ExploreProps = {
  name: string;
  items?: Category[];
};


const StyledMenu = withStyles({
  paper: {
    border: '1px solid #d3d4d5',
    maxHeight: "auto",
    width: "100%",
    maxWidth: "1100px",
    overflow: "visible",
    marginLeft: theme.spacing(88),
    marginTop: theme.spacing(4),
    padding: 0,
    "& .MuiList-padding":{
      padding:0
    }
  },
})((props: any) => (
  <Menu
  {...props}
    elevation={0}
    getContentAnchorEl={null}
    anchorOrigin={{
      vertical: "bottom",
      horizontal: "right",
    }}
    transformOrigin={{
      vertical: "top",
      horizontal: "center",
    }}

  />
));

const useStyles = makeStyles(() => ({
  button: {
    textTransform: "none",

    fontFamily: "Roboto",
    fontSize: 14,
    fontWeight: "normal",
    fontStretch: "normal",
    fontStyle: "normal",
    lineHeight: 1.57,
    letterSpacing: "normal",
    color: COLORS.CHARCOAL_GREY,

  },

  div: {
    paddingTop: 0,
  },

  root: {
    "& .MuiList-root MuiMenu-list MuiList-padding": {
      paddingTop: 0,
    },
    marginLeft: theme.spacing(2),
  },
}));

const Explore: React.FC<ExploreProps> = (props) => {

  const classes = useStyles();
  const history = useHistory();
  const dispatch = useDispatch();

  const [expand, setExpand] = useState(true);
  const [anchorEl, setAnchorEl] = useState<any>(null);

  const handleClick = (event: MouseEvent | FormEvent) => {
    if(props.name==="My Library")
    return
    setAnchorEl(event.currentTarget);

  };

  const handleClose = () => {
    setAnchorEl(null);
    setExpand(false);
  };

  const handleCategory = (category: string) => {
    console.log(category);

    dispatch({ type: "SET_CATEGORY", payload: category });
    history.push("/explore");
    handleClose();
  };


  return (
    <div className={classes.root}>
      <Button
        color="primary"
        onClick={handleClick}
        endIcon={
          !Boolean(anchorEl) ? (
            <ArrowDropDownOutlinedIcon />
          ) : (
            <ArrowDropUpOutlinedIcon />
          )
        }
        className={classes.button}
        data-testid="explore"
      >
        {" "}
        {props.name}
      </Button>


      <StyledMenu
        anchorEl={anchorEl}
        keepMounted
        open={Boolean(anchorEl)&&props.name!=="My Library"}
        onClose={handleClose}
      >
        <ExploreTab
          items={items}
          getExploreCategory={(category) => handleCategory(category)}
          handleClose={handleClose}
        />
      </StyledMenu>
    </div>
  );
};

export default Explore;

