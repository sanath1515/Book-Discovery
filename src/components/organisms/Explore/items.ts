const items =[
    {
        category:"Communication",
        topics:["Aptitude","Vocabulary","Languages","Writing"]
    },
    {
        category:"Mathematics",
        topics:["Algebra","Calculus","Statistics","Trignometry"]
    }, {
        category:"Science",
        topics:["Biology","Chemistry","Medical Science","Physics"]
    },
    {
        category:"Social Science",
        topics:["Vocabulary","Writing","Languages","Aptitude"]
    },
    {
        category:"NOVELS",
        topics:["Literary","Fiction","Genre Fiction","Science Fiction"]
    }, {
        category:"Text Books",
        topics:["Higher Education","School Textbooks","Reference","Study guides"]
    },
    {
        category:"OTHERS",
        topics:["Computer","Development","Internet","Politics"]
    },
    {
        category:"Exam Preparation",
        topics:["Defence","Engineering Entrance","Medical entrance","UPSC Exams"]
    }
  
  ]

  export default items;