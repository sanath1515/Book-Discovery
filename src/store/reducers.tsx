export type bookState = {
    searchText:string,
    category:string,
    singleViewId:string
}

export const initialBookState = {
    searchText:"",
    category:"",
    singleViewId:"1"
}

type Action = {type:string,payload:string}

export const booksReducer = (state:bookState = initialBookState,action:Action)=>
{
    switch(action.type)
    {
        case "SET_CATEGORY":{
            return {...state,category:action.payload};
        }
        case "SET_SEARCH_TEXT":{
            return {...state,searchText:action.payload};
        }
        case "SET_SINGLE_VIEW_ID":{
            return {...state,singleViewId:action.payload};
        }
        default:
            return state;
    }
}