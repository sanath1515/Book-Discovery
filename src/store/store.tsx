import { createStore } from 'redux';
import { booksReducer } from './reducers';

const store = createStore(booksReducer);

export default store;