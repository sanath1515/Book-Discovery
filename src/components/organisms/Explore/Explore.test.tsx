import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import store from "../../../store/store";
import Explore from "./Explore";

afterEach(cleanup);

describe("Explore Testing", () => {
  it("Matches Snapshot", () => {
    const { container } = render(
      <Router>
      <Provider store={store}>
      <Explore name={""} />
      </Provider>
      </Router>
    );
    expect(container).toMatchSnapshot();
  });

  it("Checking Fields", () => {
    render(
      <Router>
      <Provider store={store}>
      <Explore name={"Explore"} />
      </Provider>
      </Router>
    );

    expect(screen.getByText("Explore")).toBeInTheDocument();
  });
  it("Should click correctly", () => {
    const { getByTestId } = render(
      <Router>
        <Provider store={store}>
        <Explore name={""} />
        </Provider>
      </Router>
    );
    expect(fireEvent.click(getByTestId("explore"))).toBeTruthy();
  });
});
