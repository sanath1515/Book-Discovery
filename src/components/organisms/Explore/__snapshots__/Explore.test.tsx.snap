// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explore Testing Matches Snapshot 1`] = `
<div>
  <div
    class="makeStyles-root-3"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-1 MuiButton-textPrimary"
      data-testid="explore"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
         
        
        <span
          class="MuiButton-endIcon MuiButton-iconSizeMedium"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5H7z"
            />
          </svg>
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
